### PR TITLE
PythonOp Enhancement: Bool and Tuple[Bool] Constants, Materialize Grads, Empty Inputs, Save In Context

### PR DESCRIPTION
**Pull request by pengwa**
### PythonOp Enhancement: Bool and Tuple[Bool] Constants, Materialize Grads, Empty Inputs, Save In Context

1. Support `bool` or `Tuple[bool]` constant type in inputs.

2. Support `ctx.set_materialize_grads(True|False)`

3. Backward op can accept empty input (that don't require grad)

4. Special handling for ORT tensors are saved in context
  **Scenario**: a tensor is generated by ORT, then it might be saved for backward by `ctx.save_for_backward(tensor)`, while `tensor`'s reference count is not increased in ORT's allocation plan, so it is possible ORT release the tensor data, before backward usage.
  **Currently**: we copy every tensor before running autograd.Function.forward(), this might be a problem for cases there are many PythonOp (for example zero stage 3). 
  **Proposal**: To avoid those unnecessary copies for tensors that are not saved in context, this change introduced a `_GlobalOpKernelInfoMap`. During the kernel first run, we will anyway copy all tensors generated from ORT, and give it to torch.autograd.Function for run, then we check whether the inputs needs to be saved in context, and save the input index that needs saving in `_GlobalOpKernelInfoMap`. Then for later iterations, we just copy what is needed.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (xadupre) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:79]: No annotation for the first one?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py
Comment (xadupre) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:59]: This could introduce a perf regression. Is it possible to convert them into torch tensor using dlpack and then save them. That way, torch would be the owner of the tensors and there would not be any copy.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (xadupre) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:34]: This PR just moves this somewhere else or is it doing something more clever?

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:79]: BTW I saw your other comments, I was considering to run the forward pass once and collect all ortvalues that are saved in ctx, then for subsquent forward passes, we know exactly what tensors we need excpkictly copy before giving it to torch.autograd.Function.forward(). Doing this way, users code don't need to be aware of the copy but still be safe since we(ortmodule) copied for them.

Comment (xadupre) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:79]: I was wondering if there was a way to avoid a copy. dlapck only implies ownership transfer so it cannot be used for every saved tensor. Maybe we could save the OrtValue instead of the torch tensor but it seemed a risky change.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:79]: I am afraid, for those tensors that needs to be saved in context, we have to copy it, unless we let ORT allocation planner know there is such a reference, and know when it releases it. While knowing it requires ORT to run the forward pass once, which is behind the allocation plan generation (during session creation). ORT not ware of such a reference, then it will call torch allocator's API to release the memory, then the data got changed on that address. When backward runs, though ctx get unpack a tensor from the saved address, but the content is not expected any more.

Comment (xadupre) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:79]: I understood that's what happen when ORT receives an OrtValue through the dlpack protocol. But what it receives an OrtValue already allocated (like run_with_ortvalues). torch des not need this copy. My guess was there should be a way to avoid it too when torch is calling ort. But the time may be not significant enough to think about that.

Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: FYI I updated the code, the details is in PR description.

**Comment by pengwa** [comment]
Force pushed as a single commit, to unblock the use in local development before this PR is merged.

File: orttraining/orttraining/core/framework/torch/torch_proxy.cc
Comment (askhade) [inline orttraining/orttraining/core/framework/torch/torch_proxy.cc:226]: nit: 
```suggestion
    if (!tensor_args[i].has_value()) {
```

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3997]: nit:
```suggestion
          "Be noted: input name will be empty when its grad input is not needed by the autograd.Function.backward."
```

File: orttraining/orttraining/core/graph/training_op_defs.cc
Comment (askhade) [inline orttraining/orttraining/core/graph/training_op_defs.cc:3998]: nit:
```suggestion
          "In torch, if a forward tensor doesn't require gradient, then it's corresponding grad will be all zeros or None (depending on the setting of ctx.set_materialize_grads."
```

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:91]: nit: 
```suggestion
def _finalize_forward(
```

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:215]: done you want to check the _GloabalOpInfoKernalMap for this case?

**Review by pengwa — COMMENTED** [review]
I have adapted the code according to some suggestions. :)

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: changed to _finalize_traing_mode_forward to better explain itself. :)

**Review by askhade — APPROVED** [review]
LGTM

**Pull request merged**
